#  Basics of Python Programming Language

<h2> Python Character Set </h2>

<p> Python character set is a sequence of characters that are allowed to be used in Python Program.These characters are interpreted or uniquely identified by Python Interpreter in various contexts such as identifiers, names, characters, constants, etc. Python uses the following character set: </p>

<p>
<b>Letters:</b> Upper and lower case letters<br>
<b>Digits:</b> 0 to 9<br>
<b>Special Symbols:</b> Python supports all kinds of special symbols like, ” ‘ l ; : ! ~ @ # $ % ^ ` & * ( ) _ + – = { } [ ] \ .<br>
<b>White Spaces:</b> White spaces like tab space, blank space, newline, and carriage return.<br>
<b>Others:</b>  All ASCII and Unicode values are supported in Python Programming Language.<br>
</p>

<h2>Token</h2>

<img src="https://i.ibb.co/V3pZ54G/pt.png" alt="pt" border="0">

<h2>Literals</h2>

Literal are numbers, string, or a character that appear directly in a program. Here a Few Examples:<br><b>Examples</b><br>78    #Integer Literal<br> 21.98    #Floating literal<br>'Q'    #Character Literal<br>'Sheetal Jain'    #String Literal<br><p>We can also say a literal is a value that can be assigned to a variable. And to know the type of value python provides us an inbuilt function, <b>type()</b>. Inside type function we can pass a literal or a variable that has been assigned a particular type of literal.</p>

In [14]:
#Example 1
type(74)

int

In [15]:
#EXample 2
type("Sheetal")

str

## Identifiers 

Python Identifier is the name we give to identify a variable, list, tuple, sets, dictionary, function, class,
module or other object. That means whenever we want to give an entity a name, that’s called identifier

In [3]:
#Example 
a=10
b=20
c=a+b
print(c)

30


### Rules to define identifiers

1. We can use both upper and lower case alphabets, digits(0 to 9), underscore(_)
2. Identifier should'nt start with a digit
3. Pyhton is case sensitive i.e. total, TOTAL are two different variable
4. We can not use keyword like if, def, for, etc. as identifiers.
5. There is no length limit for python identifier. But it's not a good practice of programming
6. If the identifier starts with one underscore then it is a private identifier, and if starts with two underscores then it is strongly private.
7. If it starts as well as ends with two underscore then it is a language specific identifier defined by pyhton itself.

In [4]:
#Examples of valid identifiers
a=10
a123=45
a234b456c="Sheetal"
_add=a+10
print(a," ",a123," ",a234b456c," ",_add )

10   45   Sheetal   20


In [6]:
#Examples of invalid indentifiers
#Example 1
124abc=10

SyntaxError: invalid decimal literal (815830660.py, line 2)

In [7]:
#Example 2
def = 56

SyntaxError: invalid syntax (3331726896.py, line 2)

In [8]:
#Example of case sensitive property
total=40
TOTAL=50
print(total,"  ",TOTAL )

40    50


In [10]:
#Example of no length limit property
sssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss="Simmi sheetal jain"
print(sssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss)

Simmi sheetal jain


In [16]:
#Finding type of a varible using type function 
a=45.67
print(type(a))

<class 'float'>


<h2> Keywords </h2>

<h4>Keywords are reserved words that have a special meaning to the compiler. All keywords are written in small except True, False, and None. </h4>
<br>
Following are mostly used keywords in Python Programming
<br>
1. and, or, not (Logical operator)<br>
2. as (it is used to create an alternative name, i.e. alias)<br>
3. assert( It is used for dubugging. For example, when we do CP, cases are mathced using assert keyword)<br>
4. break( To exit from a loop)<br>
5. class (it is used to create user daefined data type in python<br>
6. coninue (Used to skip iterations)<br>
7. def ( to define a user defined function)<br>
8. del (to delete an object)<br>
9. elif ( conditional statement "else if" in python)<br>
10. if, else (It is condition statement used)<br>
11. except (to handle an exception instead of writing catch, we write except in python)<br>
12. for, while ( looping statements)<br>
13. finally ( it is used to handle exception in python)<br>
14. True, False (Boolean Value)<br>
15. from ( to import a module from a package)<br>
16. global ( it is used to declare a global variable)<br>
17. import ( to import a package)<br>
18. is (it is usedto test if two varibles are equal)<br>
19. in (to check presence of a member in an iterative object, it is also known as membership operator )<br>
20. lambda (Used to create anonymous function)<br>
21. None (equivalent to null keyword)<br>
22. pass ( pass means "do nothing". It is used to define an empty block of code)<br>
23. raise (Like in other programming language we use "throw" to throw an exception , in python we use "raise" keypword inplace of throw)<br>
24. return ( it is used to return some result (or exit) from a function)


In [ ]:
# run to get list of reserved keyword
help()


Welcome to Python 3.11's help utility!

If this is your first time using Python, you should definitely check out
the tutorial on the internet at https://docs.python.org/3.11/tutorial/.

Enter the name of any module, keyword, or topic to get help on writing
Python programs and using Python modules.  To quit this help utility and
return to the interpreter, just type "quit".

To get a list of available modules, keywords, symbols, or topics, type
"modules", "keywords", "symbols", or "topics".  Each module also comes
with a one-line summary of what it does; to list the modules whose name
or summary contain a given string such as "spam", type "modules spam".

help> keywords

Here is a list of the Python keywords.  Enter any keyword to get more help.

False               class               from                or
None                continue            global              pass
True                def                 if                  raise
and                 del                 import    

In [4]:
#Importing all keywords in python
import keyword
print(keyword.kwlist)

['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


<h2>Operators</h2>

<p>Python operator are as same as operators in other programming languages. But there are some important things to remember: </p>

#### 1. Division and Floor Division 

We use double slash for floor division. When we apply floor division, we get only number which is there before decimal. But if we want the actual quotient with decimal, we use single slash.

In [7]:
#Example 1: 
print(13000//22)
print(13000/22)

590
590.9090909090909


Though the actual value would be like this : 590.9090909090909.
But as  we are doing floor division, only 590  will be produced in output

#### 2. Membership Operator

"in" and "not in" are known as membership operator. We use them to check whether an object or element is or isn't a member of a collection or sequence. These operators return "True" or "False "

In [10]:
name="Simmisheetaljain"
print('n' in name)
print('k' in name)

True
False


#### 3. Identity Operator 

Identity Operators are used to check whether a value or expression is of certain type or class. Indentity operator returns either "True" or "False". We have two identity operators: "is" and "is not"

In [11]:
a=10
b=5.6
print( a is b)
print(a is not b)

False


#### 4. Exponent Operator (**)

We can use '**' in python to evaluate exponent values. Semantics of exponent operator are as follows:

<img src="https://i.ibb.co/cJvhvjP/IP.jpg" alt="IP" border="0">

In [12]:
#Example
print(2**5)

32


## Operator Precedence

<img src="https://i.ibb.co/0hW41wk/OP.png" alt="OP" border="0">


<img src="https://i.ibb.co/C2dXnvN/Operator-Precedence.png" alt="Operator-Precedence" border="0">

<h2>Delimiters</h2>

Delimiters are special symbols that perform a special role in Python like grouping, punctuation, and assignments. Python uses following symbols as delimiters:<br>( ) { } [ ]<br>, : . ' =  ;<br>+= -= %= //= /= *=<br>&= |= ^= >>= <<= **=

In [5]:
#print function
print("We Can Do it!!")

We Can Do it!!


In [7]:
## more about print function
print(10, end='\n')
print(20)
print(30)

10
20
30


# Separator

By default the separator for the print function is a white space. For Example:
    

In [1]:
print("Simmi","Radha","Seema","Anjali")

Simmi Radha Seema Anjali


As it can be seen, the names written in double quotes are separated by a white space.

But we can define a seprator of our own desire.

<h2>The print() function</h2>

The task of the print function is to display the content on the screen. The argument of print function can be value of any type, int, str, float, etc. It an also be a value stored in a varible.

In [4]:
#Example 1:
print("Simmi","Radha","Seema","Anjali",sep = ",")

Simmi,Radha,Seema,Anjali


In [6]:
#example 2
a=20
print("A = ", a)

A =  20


 <h2>Escape Sequence</h2>

Suppose you want to print the below line:<br><br> The ticket collector asked, "May I see your ticket?".<br><br>In python second quotation mark is considered as the end of string and it doesn't know what to do with rest of the characters.So, the string of above type will raise an error. To overcome such errors we use a special notation. This special notation consists of a backslash followed by a letter or a combination of digits and is called an <b>escape sequence</b><br> Here is a list of escape sequence used in python: <br><br><img src="https://qph.cf2.quoracdn.net/main-qimg-53fed3b1044f9500d387393497823b7e-lq">

In [18]:
#Example:
print("The ticket collector asked, \"May I see your ticket?\".")

The ticket collector asked, "May I see your ticket?".


In [19]:
#Example:
print("I\nam\nSheetal\nJain")

I
am
Sheetal
Jain


<h3>The print() function with end argument</h3>

In [20]:
#Example 
print("I am", end='\t')
print("I live in ", end='\n')
print("Simmi Sheetal Jain")

I am	I live in 
Simmi Sheetal Jain


In [21]:
#Example 
a=10
b=20
print(a,b, end='\b')

10 20

## Assigning value to a variable and then printing it:

In [27]:
#Example 1
Name="Simmi Sheetal Jain"
print("Welcome ", Name)

Welcome  Simmi Sheetal Jain


In [28]:
#Example 2
print("Welcome back ", Name, " in the world IT.")

Welcome back  Simmi Sheetal Jain  in the world IT.


In [39]:
#Example 3
_123="Sheetal"
print("My name is",_123)

My name is Sheetal


#### Assigning one value to multiple values at once 

In [33]:
#Example 1
P=Q=R=100
print(P)
print(Q)
print(R)

100
100
100


#### Assigning multiple values to multiple varibles at once

In [35]:
#Example 2
x,y,z = 38,"Sheetal Jain", 75.66
print("Roll No.:",x,"\nName: ", y,"\nPercentage: ",z)

Roll No.: 38 
Name:  Sheetal Jain 
Percentage:  75.66


## The input() Function

input() function is used to take input from the user. It takes input in string format no matter whether the input is a string or float or int.

In [40]:
#Example 
name=input("Enter your name : ")
age=input("Enter Your Age : ")
print("You Name is ", name)
print("Data type of your name is : ",type(name))
print("Your age is : ",age)
print("Data type of your age is : ",type(age))

Enter your name : Sheetal Jain
Enter Your Age : 22
You Name is  Sheetal Jain
Data type of your name is :  <class 'str'>
Your age is :  22
Data type of your age is :  <class 'str'>


## Type casting 

There are two types of type casting: Implicit and explicit. In implicit type casting, the type of a varible is changed without programmer's intervention. In Explicit type casting, programmer can change type of a varible using following functions:<br> 1. int()  #To change to integer type<br> 2. float() #to change to a float<br>3. bool() #to chnage to a boolean data type<br> 4.str() #to change to a string data type.

In [42]:
#Example 
name=input("Enter your name : ")
age=input("Enter Your Age : ")
age=int(age)
print("You Name is ", name)
print("Data type of your name is : ",type(name))
print("Your age is : ",age)
print("Data type of your age is : ",type(age))

Enter your name : Simmi
Enter Your Age : 22
You Name is  Simmi
Data type of your name is :  <class 'str'>
Your age is :  22
Data type of your age is :  <class 'int'>


## The eval() Function

"eval" means evaluation. The eval() function takes a string parameter and returns it as a Python Expression. The eval() function returns the argument passed in the type it is expected.

In [46]:
#Example
X =eval('123')
print(X)
print(type(X))

123
<class 'int'>


## Applying eval() to input()

We don't know what value the user would enter. User may enter any type of data such as a string, an integer value, a float, a complex number, etc. By making use of eval(), we can avoid type casting of the entered input in required data type. When eval() function is applied to an input function, python automatically determines the type of input entered by the user. 

In [1]:
#Example 
name=eval('input("Enter a number : ")')
if type(name)==type(23):
    print('Enter input is an integre')
else:
    print("It's not a number")

Enter a number : Simmi
It's not a number


## Formating Numbers and String

Formating a number or a string makes it presentable to a user. For example, if your want to print a floating point decimal number upto two digits after decimal, you can use format() fucntion as follows:

In [7]:
#Example 
x = 234.763544356
print("Number is : ",format(x,".2f"))

Number is :  234.76


#### Syntax of format() : 

<p>format(item, format-specifier)</p>

In [8]:
#Examples
print(format(12.34567,"10.2f"))
print(format(12,"10.2f"))
print(format(12.3,"10.2f"))

     12.35
     12.00
     12.30


### The below table shows frequently used specifiers.

 <img src="https://shorturl.at/gACIT" alt="IP" border="0"> 

### Integer Fomatting

We use d and x in integer formatting. d indicates that an integer is to be formatted and x indicates that an integer is formatted into a hexadecimal.

In [12]:
#Example 1
print(format(12,"10d"))   #Right Justification 
#Example 2
print(format(20,"10x"))   #Integer formatted to Hexadecimal Integer

        12
        14


### Formatting a String

We use s for formatting string. String, by default, is left justified. To Justify it to the left, we do formatting as follows:

In [18]:
#Example 1
print(format("Simmi Sheetal Jain","25s"))
print(format("Simmi Sheetal Jain",">25s"))
print(format("Simmi Sheetal Jain",">30s"))
print(format("Simmi Sheetal Jain",">35s"))
print(format("Simmi Sheetal Jain",">40s"))

Simmi Sheetal Jain       
       Simmi Sheetal Jain
            Simmi Sheetal Jain
                 Simmi Sheetal Jain
                      Simmi Sheetal Jain


Here, in the above function, '25s' specifies that the string will be formatted with a width of 25 and '>' indicates right justification.

### Formatting as a Percentage

We use % to format as a percentage.

In [1]:
#example 1
print(format(0.31456,"10.2"))
print(format(3.31,"10.2%"))
print(format(1.314,"10.2%"))

      0.31
   331.00%
   131.40%


The number is multiplied by 100 to format it as a percentage. In the above example "10.2%" indicates that the integer is to be formatted with a width of 10. In width, % is counted as one space.

### Formatting Scientific Notation

To format a floating number, we used "f". To format a floating point number, we can also use scientific notation using the conversion code "e".

In [23]:
#example : 
print(format(31.23746,"10.2e"))

  3.12e+01


## The ord() and chr() Functions

<b>ord()</b> function is used to return ascii value of a character passed in it as an argument. A character can be a single character or a special symbol or a number. And <b>chr()</b> is used to return the character corresponding to the value passed in it.

In [50]:
#Example 
print(ord('$'))
print(chr(698))

36
ʺ


## How to return size of a varible

In [31]:
import sys
a=45
sys.getsizeof(a)

28

## Python changes variable size dynamically

Python changes size of a varible dynamically because it is a dynamic programming language

In [30]:
#Example 
import sys
d=1287368745662776547356245552
print(d)
print(sys.getsizeof(d))
d=394888885674678678465783678467657643864785364876578346578346847474
print("New size : ")
print(sys.getsizeof(d))

1287368745662776547356245552
40
56


<h3>True and False Keyword and their working</h3>

False means 0 or lower bit, True means 1 or higher bit.

In [1]:
#Examples
False*True

0

In [2]:
False+True

1

In [3]:
True+True

2

<b>Note : </b> The only values which are false :  0, 0.0, False, 0+0j

In [35]:
#Example 
a=bool(0)
print(a)
a=bool(0.0)
print(a)
a=bool(0+0j)
print(a)
a=bool(False)
print(a)

False
False
False
False


<h2>Basics of String</h2>

In [1]:
a="Anjali"
c=" and "
b="Sheetal"
d=" will not stop till they do something great"
print(a+c+b+d)

Anjali and Sheetal will not stop till they do something great


In [7]:
msg="Sheetal's aim is to make Indian IT industry a bit more accessible"
print(msg)

Sheetal's aim is to make Indian IT industry a bit more accessible


In [8]:
msg=""" Let's learn "Python" """

In [9]:
print(msg)

 Let's learn "Python" 


In [2]:
#negativeo indexing
word="SHEETAL JAIN"
print("3rd :",word[-10])

3rd : E


In [18]:
#non-negative
word="SHEETAL JAIN"
print("3rd :",word[2])

3rd : E


In [7]:
#Contatenation
s1="Good" 
s2="Morning" 
s3=s1+" "+s2
print(s3)

Good Morning


In [5]:
m="Sheetal Jain can"
print(m,end=", ")
print("But without ",end="")
print(m[0],end="")
print(m[-2],end="")
print(m[-6],end="")
print(" , she is half of her actual strength")

Sheetal Jain can, But without Sai , she is half of her actual strength


In [4]:
m="SimmiSheetalJain"
print(m[2:12:2])

mihea


## Preactice Questions<br>
Q1. What is a variable?<br>
Q2. How to declare a varible<br>
Q3. WAP to print sum of two numbers.<br>
Q4. WAP to print area and perimeter of a rectangel <br>
Q5. WAP to print area and perimeter of a circle<br>
Q6. WAP to print type of varible <br>
Q7. WAP to print following information of a studens<br>
Name: Simmi Jain <br>
Enrollment Number: 233BCA293866<br>
Percentage: 81<br>
Section: A<br>
    

In [3]:
#Ans of Q4
l=int(input("Enter width: "))
b=int(input("Enter length: "))
print("Area is: ", l*b)
print("Perimeter is : ", 2*(l+b))

Enter width: 5
Enter length: 7
Area is:  35
Perimeter is :  24


In [4]:
#Ans of Q5
r=int(input("Enter radius: "))
print("Area is: ", 3.14*(r**2))
print("Perimeter is: ", (2*3.14*r))

Enter radius: 5
Area is:  78.5
Perimeter is:  31.400000000000002


In [5]:
#Ans of Q6
a=10
print(type(a))
a=20.0
print(type(a))
a="30"
print(type(a))
print(type(10>20))
print(type(10+34.5))

<class 'int'>
<class 'float'>
<class 'str'>
<class 'bool'>
<class 'float'>


In [9]:
#Ans of Q7
Name = "Simmi Jain"
Enrollment_Number = "233BCA293866"
Percentage = 81 
Section = 'A'
print("Name: ", Name)
print("Enrollment : ",Enrollment_Number)
print("Pecentage : ",Percentage)
print("Section: ",Section)
#Ids of all varibles 
print("\n")
print("ID of Name: ", id(Name))
print("ID of Enrollment : ",id(Enrollment_Number))
print("ID of Pecentage : ",id(Percentage))
print("ID of Section: ",id(Section))

Name:  Simmi Jain
Enrollment :  233BCA293866
Pecentage :  81
Section:  A


ID of Name:  1299643207856
ID of Enrollment :  1299642656560
ID of Pecentage :  140727251082536
ID of Section:  140727251124224


In [2]:
def verify(num1,num2):
    if num1 > num2:
        return num1
    elif num1 == num2:
        return 1
    else:
        return num2

def display(arg1,arg2):
    if(verify(arg1,arg2)==arg1):
        print("A")
    elif(verify(arg1,arg2)==1):
        print("C")
    else:
        print("B")

display(1000,3500)

B


In [3]:
num1=100
num2=200
num3=6
if(5>=num3):
    if(num1>100 or num2>150):
        print("1")
elif(num1>=100 and num2>150):
    print("2")
else:
    print("3")

2


In [5]:
for variable_1 in range(1,5,-1):
    print("executed")
print("Nothing")

Nothing


In [6]:
grade_list = ["A+","A-","B","B+","A+","A+","A-"]
for grade in grade_list:
    if grade == "A+":
        print("TPF")
    else:
        print("Not TPF")

TPF
Not TPF
Not TPF
Not TPF
TPF
TPF
Not TPF


In [7]:
number=28
for num in range(25,30):
    if(number>num):
        print(num)
    else:
        print(num)
        break

25
26
27
28


In [8]:
for num in 23, 45, 50, 65, 76, 90:
    if(num%5!=0):
        continue
    if(num%10==0):
        print(num, end=" ")
        continue
    if(num%3==0):
        print(num, end=" ")

45 50 90 

In [9]:
for number in 10,15:
    for counter in range(1,3):
        print(number*counter, end=" ")

10 20 15 30 

In [14]:
num1=16
num2=6
while(num1>=2):
    if(num1>num2):
        num1=num1/2
    else:
        print(num1)
        break

4.0


In [15]:
a=[1,2,3]
print(a.index(3))

2


In [16]:
num_list = [100.5,30.465,-1.22,20.15]
num_list.insert(1, -100.5)
num_list.pop(0)
num_list.sort()
print(num_list[0])

-100.5


In [17]:
list1=[5,10,15,20,25]
print(len(list1))
print(list1[4])
print(list1[5])
print(list1[4:5])
list1[2]=12
print(list1)
list1=list1+[8,9]

5
25


IndexError: list index out of range

In [18]:
tuple1=(5,10,15,20,25)
print(len(tuple1))
tuple1[2]=100
print(tuple1[5])
tuple1=tuple1+(8,9,"h")

5


TypeError: 'tuple' object does not support item assignment

In [19]:
message="welcome to Mysore"
word=message[-7:]
if(word=="mysore"):
    print("got it")
else:
    message=message[3:14]
    print(message)

come to Mys


In [20]:
song="JINGLE Bells jingle Bells Jingle All The Way"
song.upper()
song_words=song.split()
count=0
for word in song_words:
    if(word.startswith("jingle")):
        count=count+1
print(count)

1


In [21]:
sample_dict = {'a':1,'b':2}
sample_dict.update({'b':5, 'c':10 })
print(sample_dict.get('b'), sample_dict.get('c'))

5 10
